components/function-controller/internal/webhook/conversion_webhook.go

```diff
@@ -0,0 +1,348 @@
+	err := w.client.Get(context.Background(),
+		types.NamespacedName{
+			Name:      in.Spec.Source,
+			Namespace: in.Namespace,
+		}, repo)
+	if err != nil {
+		return errors.Wrap(err, "while getting git repository")
+	}
+	out.Spec.Source = serverlessv1alpha2.Source{
+		GitRepository: &serverlessv1alpha2.GitRepositorySource{
+			URL: repo.Spec.URL,
+
+			Repository: serverlessv1alpha2.Repository{
+				BaseDir:   in.Spec.BaseDir,
+				Reference: in.Spec.Reference,
+			},
+		},
+	}
+
+	if repo.Spec.Auth != nil {
+		out.Spec.Source.GitRepository.Auth = &serverlessv1alpha2.RepositoryAuth{
+			Type:       serverlessv1alpha2.RepositoryAuthType(repo.Spec.Auth.Type),
+			SecretName: repo.Spec.Auth.SecretName,
+		}
+	}
+	return nil
+}
+
+func (w *ConvertingWebhook) convertStatusV1Alpha1ToV1Alpha2(in *serverlessv1alpha1.FunctionStatus, out *serverlessv1alpha2.FunctionStatus) {
+	out.Repository = serverlessv1alpha2.Repository(in.Repository)
+	out.Commit = in.Commit
+	out.Runtime = serverlessv1alpha2.RuntimeExtended(in.Runtime)
+	out.RuntimeImageOverride = in.RuntimeImageOverride
+
+	out.Conditions = []serverlessv1alpha2.Condition{}
+	for _, c := range in.Conditions {
+		out.Conditions = append(out.Conditions, serverlessv1alpha2.Condition{
+			Type:               serverlessv1alpha2.ConditionType(c.Type),
+			Status:             c.Status,
+			LastTransitionTime: c.LastTransitionTime,
+			Reason:             serverlessv1alpha2.ConditionReason(c.Reason),
+			Message:            c.Message,
+		})
+	}
+}
+
+// v1alpha2 -> v1alpha1
+func (w *ConvertingWebhook) convertFunctionV1Alpha2ToV1Alpha1(src, dst runtime.Object) error {
+	in := src.(*serverlessv1alpha2.Function)
+
+	out, ok := dst.(*serverlessv1alpha1.Function)
+	if !ok {
+		dstGVK := dst.GetObjectKind().GroupVersionKind()
+		return fmt.Errorf("unsupported convert destination version %s ", dstGVK)
+	}
+
+	out.ObjectMeta = in.ObjectMeta
+
+	if err := w.convertSpecV1Alpha2ToV1Alpha1(in, out); err != nil {
+		return errors.Wrap(err, "while converting function spec from v1alpha2 to v1alpha1")
+	}
+
+	w.convertStatusV1Alpha2ToV1Alpha1(&in.Status, &out.Status)
+	return nil
+}
+
+func (w *ConvertingWebhook) convertSpecV1Alpha2ToV1Alpha1(in *serverlessv1alpha2.Function, out *serverlessv1alpha1.Function) error {
+	out.Spec.Env = in.Spec.Env
+	out.Spec.MaxReplicas = in.Spec.MaxReplicas
+	out.Spec.MinReplicas = in.Spec.MaxReplicas
+	out.Spec.Runtime = serverlessv1alpha1.Runtime(in.Spec.Runtime)
+	out.Spec.RuntimeImageOverride = in.Spec.RuntimeImageOverride
+
+	out.Spec.BuildResources = in.Spec.ResourceConfiguration.Build.Resources
+	out.Spec.Resources = in.Spec.ResourceConfiguration.Function.Resources
+
+	out.Spec.Labels = in.Spec.Template.Labels
+
+	return w.convertSourceV1Alpha2ToV1Alpha1(in, out)
+}
+
+func (w *ConvertingWebhook) convertSourceV1Alpha2ToV1Alpha1(in *serverlessv1alpha2.Function, out *serverlessv1alpha1.Function) error {
+	if in.Spec.Source.GitRepository == nil {
```

Comment:
I would check if `Inline` is not nil:
- code is clearer and you only work with inline, other types doesn't matter
- possible nil (this shouldn't happen)